### PR TITLE
Add support of AES-GCM to PKCS11 TA

### PR DESCRIPTION
**Pull request by maroueneboubakri**
Most of the code is taken from the "Secure Key Service TA" as discussed here https://github.com/OP-TEE/optee_os/issues/6406#issuecomment-1792584624
This stills under development but the aim it get feedback on code parts which will not change.

Related issue https://github.com/OP-TEE/optee_os/issues/6406

### CODE REVIEW

File: ta/pkcs11/src/token_capabilities.c [hunk 1]
```diff
@@ -231,6 +232,7 @@ const struct pkcs11_mechachism_modes token_mechanism[] = {
 	TA_MECHANISM(PKCS11_CKM_AES_ECB, CKFM_CIPHER_WRAP),
 	TA_MECHANISM(PKCS11_CKM_AES_CBC, CKFM_CIPHER_WRAP),
 	TA_MECHANISM(PKCS11_CKM_AES_CTR, CKFM_CIPHER),
+	TA_MECHANISM(PKCS11_CKM_AES_GCM, CKFM_CIPHER),
```
Comment (etienne-lms) [inline ta/pkcs11/src/token_capabilities.c:235]: This line should be added at the end of the patch series, once AES GCM is supported by the pkcs11 TA.

File: ta/pkcs11/src/processing_aes.c [hunk 2]
```diff
@@ -14,6 +15,385 @@
 #include "processing.h"
 #include "serializer.h"
 
+/*
+ * Authenticated ciphering: (CCM / GCM)
+ *
+ * As per PKCS#11, CCM/GCM decrypt shall not revealed the data until the
```
Comment (etienne-lms) [inline ta/pkcs11/src/processing_aes.c:21]: s/decrypt/decryption/
s/revealed/reveal/

File: ta/pkcs11/src/processing_aes.c [hunk 3]
```diff
@@ -14,6 +15,385 @@
...
+ * decryption is completed and the mac verified. The SKS TA must retain the
```
Comment (etienne-lms) [inline ta/pkcs11/src/processing_aes.c:22]: s/mac/MAC/
s/SKS TA/pkcs11 TA/

File: ta/pkcs11/src/processing_aes.c [hunk 4]
```diff
@@ -14,6 +15,385 @@
...
+ * ciphered data until the CCM finalization. To do so, arrays of decrypted
+ * data are allocated during AE update processing and copied into client
```
Comment (etienne-lms) [inline ta/pkcs11/src/processing_aes.c:24]: s/client/client's/

File: ta/pkcs11/src/processing_aes.c [hunk 5]
```diff
@@ -14,6 +15,385 @@
...
+ * buffer at AE finalization.
+ *
+ * As per PKCS#11, CCM/GCM decrypt expect the tag/mac data to be provided
```
Comment (etienne-lms) [inline ta/pkcs11/src/processing_aes.c:27]: s/mac/MAC/

File: ta/pkcs11/src/processing_aes.c [hunk 6]
```diff
@@ -14,6 +15,385 @@
...
+ * inside the input data for DecryptUpdate() and friends. But the DecryptFinal
```
Comment (etienne-lms) [inline ta/pkcs11/src/processing_aes.c:28]: s/DecryptUpdate/C_DecryptUpdate/
s/DecryptFinal/C_DecryptFinal/

above and below (few occurrences)

File: ta/pkcs11/src/processing_aes.c [hunk 7]
```diff
@@ -14,6 +15,385 @@
...
+ * API does not provide input data reference hence we do not know which is the
+ * last call to DecryptUpdate() where last bytes are not ciphered data but the
+ * requested tag/mac byte. To handle this, the TA saves the last input data
```
Comment (etienne-lms) [inline ta/pkcs11/src/processing_aes.c:31]: s/mac/MAC/   (should this be AAD instead of MAC?)

File: ta/pkcs11/src/processing_aes.c [hunk 8]
```diff
@@ -14,6 +15,385 @@
...
+
+	if (ctx->pending_size &&
+	    (ctx->pending_size + in_size) >= ctx->tag_byte_len) {
+		/* Process pending tag bytes that are effective data byte */
+		uint32_t len = MIN(data_len, ctx->pending_size);
+
+		res = TEE_AEUpdate(session->processing->tee_op_handle,
+				   ctx->pending_tag, len, NULL, &ct_size);
+
+		// TODO: explain this
```
Comment (etienne-lms) [inline ta/pkcs11/src/processing_aes.c:123]: Let's explain this!

File: ta/pkcs11/src/processing_aes.c [hunk 9]
```diff
@@ -14,6 +15,385 @@
...
+	rc = PKCS11_CKR_OK;
+
+bail:
+	if (rc)
+		TEE_Free(ct);
+
+	return rc;
+}
+
+static enum pkcs11_rc reveale_ae_data(struct ae_aes_context *ctx,
```
Comment (etienne-lms) [inline ta/pkcs11/src/processing_aes.c:234]: s/reveale_ae_data/reveal_ae_data/g

File: ta/pkcs11/src/processing_aes.c [hunk 10]
```diff
@@ -14,6 +15,385 @@
...
+		}
+
+		res = TEE_AEDecryptFinal(session->processing->tee_op_handle,
+					 NULL, 0, data_ptr, &data_size,
+					 ctx->pending_tag, ctx->tag_byte_len);
+
+		if (!data_size) {
+			TEE_Free(data_ptr);
+			data_ptr = NULL;
+			DMSG_RAW("Is this expected from the Core API?\n\n");
```
Comment (etienne-lms) [inline ta/pkcs11/src/processing_aes.c:313]: Please remove trailing `\n` characters, here and there.

File: ta/pkcs11/src/processing_aes.c [hunk 11]
```diff
@@ -14,6 +15,385 @@
...
+	rc = tee2pkcs_error(res);
+	if (rc)
+		goto bail;
+
+	if (data_ptr) {
+		void *tmp_ptr = NULL;
+
+		tmp_ptr = TEE_Realloc(ctx->out_data,
+				      (ctx->out_count + 1) *
+					sizeof(struct out_data_ref));
```
Comment (etienne-lms) [inline ta/pkcs11/src/processing_aes.c:330]: indentation:
```c
		tmp_ptr = TEE_Realloc(ctx->out_data,
				      (ctx->out_count + 1) *
				      sizeof(struct out_data_ref));

File: ta/pkcs11/src/processing_aes.c [hunk 12]
```diff
@@ -14,6 +15,385 @@
...
+		ctx->out_data[ctx->out_count].size = data_size;
+		ctx->out_data[ctx->out_count].data = data_ptr;
+		ctx->out_count++;
+
+		data_ptr = NULL;
+	}
+
+	rc = reveale_ae_data(ctx, out, out_size);
+
+bail:
```
Comment (etienne-lms) [inline ta/pkcs11/src/processing_aes.c:345]: please replace all `bail` branch labels with `out` branch labels.
I remember @jenswi-linaro asked me this at the time when we reviewed the pkcs11 TA implementation.

**Comment by maroueneboubakri** [comment]
@etienne-lms please review this PR when have time.

Some on-going issues:

When calling `C_Encrypt()` to determine the output buffer size returns correctly the expected buffer size. However, the operation terminate with OK status, therefore, it terminates the operation. Next call to `C_Encrypt()` to effecively encrypt data returns `OPERATION_NOT_INITIALIZED` and this because the previous call terminated the operation (succeeded ?).

When calling `C_Decrypt()` to same gather the needed output buffer size, it returns `BUFFER_TOO_SMALL` with the correct expected output buffer size. 

Calling `C_EncryptInit()` before each  `C_Encrypt()` does not generate an error first `C_Encrypt()` succeeds while should not ?
However, Calling `C_DecyptInit()` before each  `C_Decrypt()` generates `OPERATION_ACTIVE`. Expected but, during the first call to `C_Decrypt()` the context has been cleared and freeed. Therefore, `TEE_AEUpdate()` fails with `BAD_PARAMS` (Tag buffer has been already cleared by the first call).

Coudl please help here, seems that flow has been considerably changed since first time the code has been introduced.

Thank you
Maro

File: ta/pkcs11/src/processing_aes.c [hunk 13]
```diff
@@ -50,3 +430,112 @@ enum pkcs11_rc tee_init_ctr_operation(struct active_processing *processing,
...
+		return PKCS11_CKR_ARGUMENTS_BAD;
+
+	serialargs_init(&args, proc_params, params_size);
+
+	rc = serialargs_get(&args, &iv_len, sizeof(uint32_t));
+	if (rc)
+		goto bail;
+
+	// TODO: no need to copy iv into secure world
+	rc = serialargs_alloc_and_get(&args, &iv, iv_len);
```
Comment (etienne-lms) [inline ta/pkcs11/src/processing_aes.c:460]: could replace with `serialargs_get_ptr()`
Ditto for AAD's at line 469.

File: ta/pkcs11/src/processing_aes.c [hunk 14]
```diff
@@ -50,3 +430,112 @@ enum pkcs11_rc tee_init_ctr_operation(struct active_processing *processing,
...
+	if (rc)
+		goto bail;
+
+	tag_len = ROUNDUP(tag_bitlen, 8) / 8;
+
+	/* As per pkcs#11 mechanism specification */
+	if (tag_bitlen > 128 ||
+	    !iv_len || iv_len > 256) {
+		DMSG("Invalid parameters: tag_bit_len %" PRIu32
+			", iv_len %" PRIu32, tag_bitlen, iv_len);
```
Comment (etienne-lms) [inline ta/pkcs11/src/processing_aes.c:483]: *indentation

**Review by etienne-lms — COMMENTED** [review]
Commit "ta: pkcs11: pkcs11_ta.h: define PKCS11_CKM_AES_GCM":
s/0x01086/0x01087/ in the commit message.
With that fixed:  `Reviewed-by: Etienne Carriere <etienne.carriere@foss.st.com>`

Commit "ta: pkcs11: pkcs11_attributes.c: set attributes for AES-GCM":
Please add  (with a `.` in the message message to end the sentence).
With that fixed:  `Reviewed-by: Etienne Carriere <etienne.carriere@foss.st.com>`.

Commits "ta: pkcs11: processing.h: add prototypes of authenticated encryption" should be removed (IMO) and prototypes be declared in the very same commit that implements the function.

Commit "ta: pkcs11: processing.c: release operation of AES-GCM":
`Reviewed-by: Etienne Carriere <etienne.carriere@foss.st.com>`.

**Comment by etienne-lms** [comment]
@maroueneboubakri: if you're to make changes in this series, could you append fixup commits on top the existing tree until reviewers ask them to be squashed? It will ease understand how the series change since last review pass.

**Comment by etienne-lms** [comment]
Hi @maroueneboubakri, I posted some comments right after you posted https://github.com/OP-TEE/optee_os/pull/6467#issuecomment-1824629664. Be sure i'll take time to give you feedback (possibly answer your questions if a can) as soon as I can.

File: ta/pkcs11/src/processing_aes.c (hunk 7 above)
Comment (maroueneboubakri) [reply ta/pkcs11/src/processing_aes.c:31]: I believe it is MAC.

File: ta/pkcs11/src/processing_aes.c (hunk 8 above)
Comment (maroueneboubakri) [reply ta/pkcs11/src/processing_aes.c:123]: It should be somehow related to to my concern above https://github.com/OP-TEE/optee_os/pull/6467#issuecomment-1824629664

**Comment by maroueneboubakri** [comment]
> Hi @maroueneboubakri, I posted some comments right after you posted [#6467 (comment)](https://github.com/OP-TEE/optee_os/pull/6467#issuecomment-1824629664). Be sure i'll take time to give you feedback (possibly answer your questions if a can) as soon as I can.

Thanks @etienne-lms ! Comments above addressed.

**Comment by maroueneboubakri** [comment]
@etienne-lms Fixed applied tags and commits order.

**Comment by maroueneboubakri** [comment]
@etienne-lms can we check this PR ? mainly the issue of operation termination when client request output buffer size for Encrypt and for Decrypt the whole context is cleared when requesting buffer size.

**Comment by maroueneboubakri** [comment]
@etienne-lms hope you will have time to review this, for one shot operation stills needs to invoque `TEE_AEUpdate()` to get the output buffer size.

**Comment by maroueneboubakri** [comment]
Here is the issue: When we call `C_Decrypt()` with output beffer is NULL to determine output buffer length, `tee_ae_decrypt_final()` returns `PKCS11_CKR_BUFFER_TOO_SMALL `, however, the operation state is no longer active, due to call `TEE_AEDecryptFinal()` inside `tee_ae_decrypt_final()`. So next call to `C_Decrypt()` with output buffer allocates and good output size will make call to `TEE_AEUpdate()` `through tee_ae_decrypt_update()` which panics with error code `TEE_ERROR_BAD_PARAMETERS` since operation is nolonger valid.

**Comment by etienne-lms** [comment]
Hello @maroueneboubakri,
As per GP TEE spec, a call to `TEE_AEDecryptFinal()` that gets `TEE_ERROR_SHORT_BUFFER` keeps operation in **active** state. I think OP-TEE's implementation conforms to that but i'll check.
> 6.6.5 TEE_AEDecryptFinal
> (...)
> The operation may be in either **initial** or **active** state. If the result is not TEE_ERROR_SHORT_BUFFER , the
operation enters **initial** state afterwards.

**Comment by maroueneboubakri** [comment]
> TEE_AEDecryptFinal

@etienne-lms When `TEE_AEDecryptFinal()` returns short buffer, `tee_ae_decrypt_final()` allocates necessary memory and calls again `TEE_AEDecryptFinal()` which then succeeds. But `tee_ae_decrypt_final()` returns short buffer in the end since the initial ouput size is compared with the result from the last call to `TEE_AEDecryptFinal()`.

File: ta/pkcs11/src/processing_symm.c [hunk 15]
```diff
@@ -829,7 +838,39 @@ enum pkcs11_rc step_symm_operation(struct pkcs11_session *session,
...
+				 rc == PKCS11_CKR_BUFFER_TOO_SMALL)) {
+				out_buf = (char *)out_buf + out_size;
+				tmp_size -= out_size;
+			}
+			break;
+		case PKCS11_FUNCTION_DECRYPT:
+			rc = tee_ae_decrypt_update(session, in_buf, in_size);
+			/* Keep decrypted data in secure memory until final */
+			out_size = 0;
+			output_data = true;
```
Comment (etienne-lms) [inline ta/pkcs11/src/processing_symm.c:867]:
```c
			assert(rc != PKCS11_CKR_SMALL_BUFFER);
			if (rc)
				return rc;

File: ta/pkcs11/src/processing_symm.c [hunk 16]
```diff
@@ -829,7 +838,39 @@ enum pkcs11_rc step_symm_operation(struct pkcs11_session *session,
...
+			res = TEE_AEUpdate(proc->tee_op_handle,
+					   in_buf, in_size, out_buf, &out_size);
+			output_data = true;
+			rc = tee2pkcs_error(res);
+			if (step == PKCS11_FUNC_STEP_ONESHOT &&
+			    (rc == PKCS11_CKR_OK ||
+				 rc == PKCS11_CKR_BUFFER_TOO_SMALL)) {
+				out_buf = (char *)out_buf + out_size;
+				tmp_size -= out_size;
+			}
```
Comment (etienne-lms) [inline ta/pkcs11/src/processing_symm.c:861]:
```c
			if (rc && rc != PKCS11_CKR_SMALL_BUFFER)
				return rc;

**Comment by maroueneboubakri** [comment]
Comments addressed.

**Comment by etienne-lms** [comment]
Still in progress, please keep alive.

**Comment by maroueneboubakri** [comment]
Closed since superseded by https://github.com/OP-TEE/optee_os/pull/6683.

**Pull request closed without merging**